### DISCUSSION TITLE
Does limiting battery charge work?

### DISCUSSION THREAD

**Discussion by h1v9** [kuruczgy/x1e-nixos-config#41] [Q&A] [answered]
Just like the title says. On windows it's really not possible to limit battery charge via software, is it achieavable on linux? AFAIK you should be able to do it with tlp or upower

**Comment by kuruczgy** [comment]
Short answer: I don't know, probably not yet.

The way I understand, a coprocessor inside the SoC is responsible for battery management. From https://github.com/jhovold/linux/blob/wip/x1e80100-6.12-rc7/Documentation/devicetree/bindings/soc/qcom/qcom%2Cpmic-glink.yaml:
> The PMIC GLINK service, running on a coprocessor on some modern Qualcomm platforms and implement USB Type-C handling and **battery management**.

Here is the driver: https://github.com/jhovold/linux/blob/wip/x1e80100-6.12-rc7/drivers/power/supply/qcom_battmgr.c
Skimming through it, I cannot find anything related to charge limiting. This could mean either that
- it is just not implemented yet in the driver, but the PMIC GLINK service supports it,
- or it could mean that there is no support for it in the service. Unfortunately the code running on the coprocessor is closed source (and probably signed so there is no way for us to modify it), so if there is no support then we have to wait for qcom to provide it.

You could also ask in #aarch64-laptops (you can join on OFTC or over the matrix bridge, see here: https://github.com/aarch64-laptops/build#contact-us), there are some much more knowledgeable people there. (Also it seems that the battery management is mostly identical to the one of the sc8280xp, which is also an SoC with a lot of community knowledge about it.)

**Comment by kuruczgy** [comment, marked as answer]
Supported with https://github.com/kuruczgy/x1e-nixos-config/pull/140
@h1v9 if you are still curious
and yes, it's still not possible on Windows AFAICT, so now for this tiny thing Linux actually has better hardware support :)
